modules/payment/src/Plugin/Commerce/CheckoutPane/PaymentProcess.php

```diff
@@ -133,38 +133,43 @@ public function buildPaneForm(array $pane_form, FormStateInterface $form_state,
     $payment_gateway = $this->order->payment_gateway->entity;
     $payment_gateway_plugin = $payment_gateway->getPlugin();
 
+    $payment_storage = $this->entityTypeManager->getStorage('commerce_payment');
+    $payment = $payment_storage->create([
+      'state' => 'new',
+      'amount' => $this->order->getTotalPrice(),
+      'payment_gateway' => $payment_gateway->id(),
+      'order_id' => $this->order->id(),
+    ]);
+
     if ($payment_gateway_plugin instanceof OnsitePaymentGatewayInterface) {
       try {
-        $payment_storage = $this->entityTypeManager->getStorage('commerce_payment');
-        $payment = $payment_storage->create([
-          'state' => 'new',
-          'amount' => $this->order->getTotalPrice(),
-          'payment_gateway' => $payment_gateway->id(),
-          'payment_method' => $this->order->payment_method->entity,
-          'order_id' => $this->order->id(),
-        ]);
+        $payment->payment_method = $this->order->payment_method->entity;
         $payment_gateway_plugin->createPayment($payment, $this->configuration['capture']);
-
-        $next_step_id = $this->checkoutFlow->getNextStepId();
-        // @todo Add a checkout flow method for completing checkout.
-        if ($next_step_id == 'complete') {
-          $transition = $this->order->getState()->getWorkflow()->getTransition('place');
-          $this->order->getState()->applyTransition($transition);
-        }
-        $this->order->checkout_step = $next_step_id;
-        $this->order->save();
-
-        throw new NeedsRedirectException(Url::fromRoute('commerce_checkout.form', [
-          'commerce_order' => $this->order->id(),
-          'step' => $next_step_id,
-        ])->toString());
+        $this->checkoutFlow->redirectToStep($this->checkoutFlow->getNextStepId());
 
       }
       catch (PaymentGatewayException $e) {
         drupal_set_message($e->getMessage(), 'error');
         $this->redirectToPreviousStep();
       }
     }
+    elseif ($payment_gateway_plugin instanceof OffsitePaymentGatewayInterface) {
+      if ($payment_gateway_plugin instanceof OffsiteRedirectPaymentGatewayInterface) {
+        $redirect_url = $payment_gateway_plugin->getRedirectUrl();
+        $pane_form['offsite_payment']['#prefix'] = '<div class="payment-redirect-form" data-redirect-url="' . $redirect_url . '">';
```

Comment:
In most cases a direct header-based redirect would be sufficient, no? Gateways that need an HTML redirect are a good demonstration of why we need an "offsite" plugin form, this code should definitely be overridable.